src/com/google/javascript/jscomp/ConvertChunksToESModules.java

```diff
@@ -0,0 +1,297 @@
+   */
+  ConvertChunksToESModules(AbstractCompiler compiler) {
+    this.compiler = compiler;
+  }
+
+  @Override
+  public void process(Node externs, Node root) {
+    // Find global names that are used in more than one chunk. Those that
+    // are have to have import and export statements added.
+    NodeTraversal.traverse(compiler, root, new FindCrossChunkReferences());
+
+    // Force every output chunk to parse as an ES Module. If a chunk has no imports and
+    // no exports, add an empty export list to generate an empty export statement:
+    // example: export {};
+    for (JSModule chunk : compiler.getModuleGraph().getAllModules()) {
+      if (!crossChunkExports.containsKey(chunk) &&
+          !crossChunkImports.containsKey(chunk) &&
+          chunk.getInputs().size() > 0) {
+        crossChunkExports.put(chunk, new HashSet<>());
+      }
+    }
+
+    ConvertChunkSourcesToModules();
+    AddExportStatements();
+    AddImportStatements();
+  }
+
+  /**
+   * Move all code in a chunk into the first input and mark it as an ESModule.
+   * At this point in the compilation, all input files should be scripts.
+   */
+  private void ConvertChunkSourcesToModules() {
+    for (JSModule chunk : compiler.getModuleGraph().getAllModules()) {
+      if (chunk.getInputs().size() == 0) {
+        continue;
+      }
+
+      CompilerInput firstInput = null;
+      for (CompilerInput input : chunk.getInputs()) {
+        Node astRoot = input.getAstRoot(compiler);
+        FeatureSet scriptFeatures = NodeUtil.getFeatureSetOfScript(astRoot);
+        checkState(!scriptFeatures.contains(FeatureSet.ES6_MODULES));
+        if (firstInput == null) {
+          firstInput = input;
+          scriptFeatures = scriptFeatures.union(FeatureSet.ES6_MODULES);
+          astRoot.putProp(Node.FEATURE_SET, scriptFeatures);
+          Node moduleBody = new Node(Token.MODULE_BODY);
+          moduleBody.useSourceInfoFrom(astRoot);
+          moduleBody.addChildrenToFront(astRoot.removeChildren());
+          astRoot.addChildToFront(moduleBody);
+          compiler.reportChangeToEnclosingScope(moduleBody);
+        } else {
+          Node firstInputAstRoot = firstInput.getAstRoot(compiler);
+          FeatureSet firstInputScriptFeatures = NodeUtil.getFeatureSetOfScript(firstInputAstRoot);
+          FeatureSet combinedFeatureSet =
+              firstInputScriptFeatures.union(NodeUtil.getFeatureSetOfScript(astRoot));
+          astRoot.putProp(Node.FEATURE_SET, combinedFeatureSet);
+          Node moduleBody = firstInputAstRoot.getFirstChild();
+          checkState(moduleBody != null && moduleBody.isModuleBody());
+          moduleBody.addChildrenToBack(astRoot.removeChildren());
+          compiler.reportChangeToEnclosingScope(firstInputAstRoot);
+          compiler.reportChangeToChangeScope(astRoot);
+        }
+      }
+    }
+  }
+
+  /**
+   * Add export statements to chunks
+   */
+  private void AddExportStatements() {
```

Comment:
Could you add a code comment to show what you expect the code generated by this to look like?

e.g.

```java
// Append one big export statement to the end of the whole chunk.
// ```javascript
// export { name1, name2, name3, ...., nameN };
// ```
```

Reply:
done